plutus-metatheory/src/VerifiedCompilation/UntypedTranslation.lagda.md

```diff
@@ -26,168 +26,221 @@ open import VerifiedCompilation.Equality using (DecEq; _≟_; decPointwise)
 ```
 The generic type of a Translation is that it matches one (or more) patterns on the left to one
 (or more) patterns on the right. If there are decision procedures to identify those patterns,
-we can build a decision procedure to apply them recursivley down the AST structure. 
+we can build a decision procedure to apply them recursivley down the AST structure.
 
 ```
-Relation = { X : Set } → (X ⊢) → (X ⊢) → Set₁
+Relation = { X : Set } → {{_ : DecEq X}} → (X ⊢) → (X ⊢) → Set₁
 
-data Translation (R : Relation) : { X : Set } → (X ⊢) → (X ⊢) → Set₁ where
-  istranslation : { X : Set } → (ast ast' : X ⊢) → R ast ast' → Translation R ast ast'
-  var : { X : Set } → {x : X} → Translation R (` x) (` x) -- We assume we won't want to translate variables individually?
-  ƛ   : { X : Set } → {x x' : Maybe X ⊢}
+data Translation (R : Relation) { X : Set } {{_ : DecEq X}} : (X ⊢) → (X ⊢) → Set₁ where
```

Comment:
I'm not really confident that we need that `DecEq X` here or that we need it all, maybe there's some way to avoid adding it that I failed to find.

Reply:
> maybe there's some way to avoid adding it that I failed to find

Actually, we can probably just compute `X` from `n : Nat` and that's it. I.e. do this (I didn't try type checking it, not sure if the last line is gonna work for example):

```agda
NatToSet : ℕ → Set
NatToSet zero    = ⊥
NatToSet (suc n) = Maybe (NatToSet n)

Relation = ∀ n → NatToSet n ⊢ → NatToSet n ⊢ → Set₁

data Translation (R : Relation) : Relation
```

I'm not sure if it's more pain or less than the current solution though, could be either way.

Reply:
I think, at best, this is adding a layer of indirect to the current complexity, without actually removing it... You will have to start littering the code with explicit `n`s because I suspect Agda won't figure them out for itself. At worst, you will lose the quite elegant typechecking of scopes that James has made with the `Maybe`s (even if they are annoying). What you suggestion *does* achieve though is making it explicit that everything in the stack is either `Maybe` or `⊥`, and that should make the DecEq problem go away?

Reply:
> explicit `n`s because I suspect Agda won't figure them out for itself

I made it explicit, but now that I look at that definition, I think Agda should be able to figure it out, because `NatToSet` is constructor-headed and `⊢` is injective as well (from the unifier's POV, not theory).

> At worst, you will lose the quite elegant typechecking of scopes that James has made with the Maybes (even if they are annoying).

Not sure what you mean.

> What you suggestion does achieve though is making it explicit that everything in the stack is either `Maybe` or `⊥`, and that should make the DecEq problem go away?

Yup, you won't need `DecEq` if you can simply match on `n` and tell if it's another `Maybe` inside or the final `⊥`.

It may be easier in the sense that you won't need to deal with instance arguments, which were always a pain for me, although it's been a while since I last used them, so maybe they're fine these days.